File: pkg/EarthSciAST.jl/src/classification.jl
# Derived classification of a model's variables — esm-spec §6.3.1.
#
# esm 1.0.0 declares TWO variable types, `unknown` and `parameter`, and nothing
# else. Every finer category a solver needs is DERIVED from the equations and
# from the parameters' `distribution` / `update`, never declared. This file is
# the one place that derivation lives; every site that used to branch on
# `variable.type == "state"` / `"observed"` / `"brownian"` / `"discrete"` calls
# into it instead.
#
# The functions here are the cross-binding contract: all five bindings expose
# the same set, spelled in their own idiom, and
# `tests/conformance/classification/` pins one answer for all of them.

"""
    unknown_names(model::Model) -> Vector{String}

Every `unknown` declared by `model`, sorted lexicographically. The union of
[`ode_states`](@ref), [`observed_unknowns`](@ref) and
[`algebraic_unknowns`](@ref), which partition it.
"""
unknown_names(model::Model)::Vector{String} =
    sort!(String[n for (n, v) in model.variables if v.type == UnknownVariable])

"""
    parameter_names(model::Model) -> Vector{String}

Every `parameter` declared by `model`, sorted lexicographically. The union of
[`brownian_parameters`](@ref), [`discrete_parameters`](@ref),
[`sampled_parameters`](@ref) and [`constant_parameters`](@ref), which partition
it.
"""
parameter_names(model::Model)::Vector{String} =
    sort!(String[n for (n, v) in model.variables if v.type == ParameterVariable])

# ── Equation-LHS shapes ─────────────────────────────────────────────────────
#
# An equation says what it defines through the SHAPE of its left-hand side, and
# the corpus spells "define u" in several nested shapes. `_lhs_role` reduces
# them to three answers:
#
# | LHS | role |
# |---|---|
# | `D(u, t)`, `D(u[i])`, `aggregate{expr: D(index(u,i))}` | `(:derivative, "u")` |
# | `"y"`, `index(y,i)`, `aggregate{expr: index(y,i)}` | `(:definition, "y")` |
# | `ic(u)` | `(:none, "")` — an initial condition defines no dynamics |
# | `H*H*SO4`, `laplacian(phi)`, anything else | `(:implicit, "")` |

"""Strip the array-addressing wrappers (`index` / `arrayop`) and the
vectorising `aggregate` from an LHS, returning the node they address."""
function _lhs_unwrap(e::ASTExpr)::ASTExpr
    while e isa OpExpr
        if e.op == "aggregate" && e.expr_body !== nothing
            e = e.expr_body
        elseif (e.op == "index" || e.op == "arrayop") && !isempty(e.args)
            e = e.args[1]
        else
            return e
        end
    end
    return e
end

"""True iff `e` is a STRUCTURAL time derivative — a `D` whose `wrt` is `"t"` or
absent (esm-spec §4.2: an absent `wrt` means `t`)."""
_is_time_derivative(e::ASTExpr)::Bool =
    e isa OpExpr && e.op == "D" && (e.wrt === nothing || e.wrt == "t")

"""True iff `e` is a SPATIAL derivative — a `D` whose `wrt` is present and is
not `"t"`, or one of the `grad` / `div` / `laplacian` sugar ops (esm-spec
§6.3.1). Both spellings count: they are open-tier rewrite targets, so neither
is canonical."""
_is_spatial_derivative(e::ASTExpr)::Bool =
    e isa OpExpr && ((e.op == "D" && e.wrt !== nothing && e.wrt != "t") ||
                     e.op == "grad" || e.op == "div" || e.op == "laplacian")

"""
    _lhs_role(lhs::ASTExpr) -> Tuple{Symbol,String}

The role an equation's LHS plays in classification: `(:derivative, base)`,
`(:definition, base)`, `(:none, "")` for an `ic` LHS, or `(:implicit, "")` for
an expression LHS that constrains its operands only implicitly.
"""
function _lhs_role(lhs::ASTExpr)::Tuple{Symbol,String}
    head = _lhs_unwrap(lhs)
    # A DOTTED name is not rejected here. In a FLATTENED system every variable
    # key is `<ModelPath>.<name>`, so rejecting dots would leave a flattened
    # model with no states and no observeds at all. Whether the name belongs to
    # this model is decided by the caller's `model.variables` lookup, which also
    # filters out a genuinely cross-system target in an unflattened document.
    head isa VarExpr && return (:definition, head.name)
    head isa OpExpr || return (:implicit, "")
    head.op == "ic" && return (:none, "")
    if _is_time_derivative(head) && !isempty(head.args)
        base = _lhs_unwrap(head.args[1])
        base isa VarExpr && return (:derivative, base.name)
        return (:implicit, "")
    end
    return (:implicit, "")
end

# ── The three unknown sets (they partition) ─────────────────────────────────

"""
    ode_states(model::Model) -> Vector{String}

The unknowns appearing under `D(·, t)` on some equation LHS — the quantities
the solver integrates (esm-spec §6.3.1). Sorted lexicographically.

A derivative LHS may be wrapped: `D(u)`, `D(u[i])`, and an `aggregate` whose
body is a `D(…)` all credit the base variable.

This replaces every 0.x `variable.type == "state"` test.
"""
function ode_states(model::Model)::Vector{String}
    out = Set{String}()
    for eq in model.equations
        role, name = _lhs_role(eq.lhs)
        role === :derivative || continue
        v = get(model.variables, name, nothing)
        v !== nothing && v.type == UnknownVariable && push!(out, name)
    end
    return sort!(collect(out))
end

"""
    is_ode_state(model::Model, name::AbstractString) -> Bool

Membership test for [`ode_states`](@ref) — the direct replacement for a 0.x
`variable.type == "state"` branch.
"""
function is_ode_state(model::Model, name::AbstractString)::Bool
    v = get(model.variables, String(name), nothing)
    (v !== nothing && v.type == UnknownVariable) || return false
    for eq in model.equations
        role, n = _lhs_role(eq.lhs)
        role === :derivative && n == name && return true
    end
    return false
end

"""
    observed_definitions(model::Model) -> Dict{String,ASTExpr}

Each observed unknown mapped to its DEFINING EQUATION's right-hand side.

This is where the 1.0.0 relocation lands: an observed's definition used to live
in `variables[v].expression` and now lives in the model's `equations` array, as
the bare-variable-LHS equation whose LHS names it. Every site that read
`var.expression` for an observed reads this instead.

An unknown that is also an ODE state is excluded (a `D(y)` equation defines
`y`'s dynamics, not `y` itself). Where more than one equation has the same
bare-variable LHS the FIRST wins, matching the reference implementation.
"""
function observed_definitions(model::Model)::Dict{String,ASTExpr}
    states = Set(ode_states(model))
    defs = Dict{String,ASTExpr}()
    for eq in model.equations
        role, name = _lhs_role(eq.lhs)
        role === :definition || continue
        name in states && continue
        v = get(model.variables, name, nothing)
        (v !== nothing && v.type == UnknownVariable) || continue
        haskey(defs, name) || (defs[name] = eq.rhs)
    end
    return defs
end

"""
    observed_definition(model::Model, name::AbstractString) -> Union{ASTExpr,Nothing}

The defining right-hand side of the observed unknown `name`, or `nothing` when
it is not an observed unknown of `model`. Convenience over
[`observed_definitions`](@ref) for a single lookup.
"""
observed_definition(model::Model, name::AbstractString) =
    get(observed_definitions(model), String(name), nothing)

"""
    observed_unknowns(model::Model) -> Vector{String}

The unknowns defined by a bare-variable LHS (`y ~ f(…)`) — eliminable and
materializable (esm-spec §6.3.1). Sorted lexicographically.

This replaces every 0.x `variable.type == "observed"` test.
"""
observed_unknowns(model::Model)::Vector{String} =
    sort!(collect(keys(observed_definitions(model))))

"""
    algebraic_unknowns(model::Model) -> Vector{String}

The unknowns constrained only implicitly (`H*H*SO4 ~ Ksp`) — those an equation
never names as a derivative target or a bare-variable LHS (esm-spec §6.3.1).
Sorted lexicographically.

Defined as the complement so that [`ode_states`](@ref),
[`observed_unknowns`](@ref) and `algebraic_unknowns` PARTITION the unknowns by
construction: an unknown with no equation of its own lands here rather than
vanishing (the equation balance, not the classification, is what reports it).
"""
function algebraic_unknowns(model::Model)::Vector{String}
    accounted = Set(ode_states(model))
    union!(accounted, keys(observed_definitions(model)))
    return String[n for n in unknown_names(model) if !(n in accounted)]
end

"""
    solver_unknowns(model::Model) -> Vector{String}

The unknowns that occupy a SOLVER SLOT — [`ode_states`](@ref) plus
[`algebraic_unknowns`](@ref), sorted lexicographically. Equivalently, the
unknowns minus the observed ones, which are eliminable.

This is the set the 0.x declared type `state` named, and the set every
partition, cell-enumeration, and live-state guard in the tree-walk build wants:
an algebraic unknown is solved for exactly like an ODE state (it just has no
time derivative), so treating only `ode_states` as "the states" drops a
declared, array-shaped algebraic unknown out of the partition it belongs to.
It is also the CONTINUOUS half of the cadence leaf-seed table
(CONFORMANCE_SPEC §5.7.2), which seeds both alike.
"""
function solver_unknowns(model::Model)::Vector{String}
    observed = Set(observed_unknowns(model))
    return String[n for n in unknown_names(model) if !(n in observed)]
end

# ── The four parameter sets (they partition) ────────────────────────────────

"""
    _update_rules(v::ModelVariable) -> Vector{ParameterUpdate}

The variable's update rules, empty when it declares none. `update` is stored as
a vector whatever its wire spelling, so a single rule and an ordered array read
the same here.
"""
_update_rules(v::ModelVariable)::Vector{ParameterUpdate} =
    v.update === nothing ? ParameterUpdate[] : v.update

"""True iff any of the parameter's update rules is a `wiener` process. The
schema forbids `wiener` inside an array, so in practice an array update is
never Brownian — but the test reads through the vector rather than assuming
it, so a schema relaxation cannot silently change the answer."""
_is_wiener(v::ModelVariable)::Bool = any(r -> r.kind == "wiener", _update_rules(v))

"""
    brownian_parameters(model::Model) -> Vector{String}

Parameters whose `update.kind` is `"wiener"` — the SDE noise sources
(esm-spec §6.3.1). Sorted lexicographically. Their presence is what makes
[`system_kind`](@ref) `"sde"`.

This replaces every 0.x `variable.type == "brownian"` test.
"""
brownian_parameters(model::Model)::Vector{String} =
    sort!(String[n for (n, v) in model.variables
                 if v.type == ParameterVariable && _is_wiener(v)])

"""
    discrete_parameters(model::Model) -> Vector{String}

Parameters carrying any update OTHER than `wiener` — piecewise-constant between
refreshes (esm-spec §6.3.1). Sorted lexicographically. These are the sole seed
of the DISCRETE cadence class (CONFORMANCE_SPEC §5.7.2).

This replaces every 0.x `variable.type == "discrete"` test AND the 0.x
`DiscreteEvent.discrete_parameters` list.
"""
discrete_parameters(model::Model)::Vector{String} =
    sort!(String[n for (n, v) in model.variables
                 if v.type == ParameterVariable && !isempty(_update_rules(v)) &&
                    !_is_wiener(v)])

"""
    sampled_parameters(model::Model) -> Vector{String}

Parameters with a `distribution` and no `update` — drawn ONCE at setup
(uncertainty quantification, ensembles). Sorted lexicographically.
"""
sampled_parameters(model::Model)::Vector{String} =
    sort!(String[n for (n, v) in model.variables
                 if v.type == ParameterVariable && v.distribution !== nothing &&
                    isempty(_update_rules(v))])

"""
    constant_parameters(model::Model) -> Vector{String}

Parameters with neither a `distribution` nor an `update` — plain constants.
Sorted lexicographically.
"""
constant_parameters(model::Model)::Vector{String} =
    sort!(String[n for (n, v) in model.variables
                 if v.type == ParameterVariable && v.distribution === nothing &&
                    isempty(_update_rules(v))])

# ── System kind ─────────────────────────────────────────────────────────────

"""True iff any node of `e` (descending EVERY expression-bearing field, not
just `args`) satisfies `pred`."""
function _any_subexpr(pred, e::ASTExpr)::Bool
    pred(e) && return true
    for c in child_exprs(e)
        _any_subexpr(pred, c) && return true
    end
    return false
end

_equation_has(pred, eq::Equation)::Bool =
    _any_subexpr(pred, eq.lhs) || _any_subexpr(pred, eq.rhs)

"""
    has_spatial_derivative(model::Model) -> Bool

True iff some equation of `model` contains a spatial derivative anywhere in its
LHS or RHS — a `D` whose `wrt` is present and is not `"t"`, or a `grad` / `div`
/ `laplacian` sugar op (esm-spec §6.3.1).
"""
has_spatial_derivative(model::Model)::Bool =
    any(eq -> _equation_has(_is_spatial_derivative, eq), model.equations)

"""
    has_time_derivative(model::Model) -> Bool

True iff some equation of `model` contains a structural time derivative
(`D` with `wrt` `"t"` or absent) anywhere in its LHS or RHS.
"""
has_time_derivative(model::Model)::Bool =
    any(eq -> _equation_has(_is_time_derivative, eq), model.equations)

"""
    system_kind(model::Model) -> String

Derive what the `system_kind` field declares (esm-spec §6.3.1). Four conditions
tested **in this order**, first match winning:

| # | Condition | Result |
|---|---|---|
| 1 | any parameter in [`brownian_parameters`](@ref) | `"sde"` |
| 2 | any equation contains a spatial derivative | `"pde"` |
| 3 | no time-derivative equation at all | `"nonlinear"` |
| 4 | otherwise | `"ode"` |

Both orderings are load-bearing. `"pde"` precedes `"nonlinear"` so a
steady-state PDE (`laplacian(phi) ~ f`, no ∂/∂t) is a PDE rather than falling
through to `"nonlinear"`. `"sde"` precedes `"pde"` because there is no
`SPDESystem` constructor to select, so a model that is both assembles as an SDE.

The rule is stated over the EQUATIONS and not over a `domain` block: v0.8.0
removed `Domain.spatial`, so `domain` says nothing about space and a
domain-based test is not implementable.

A binding uses this derivation when the `system_kind` field is absent, and
reports `system_kind_mismatch` when a present field contradicts it — see
[`declared_system_kind_mismatch`](@ref).
"""
function system_kind(model::Model)::String
    isempty(brownian_parameters(model)) || return "sde"
    has_spatial_derivative(model) && return "pde"
    has_time_derivative(model) || return "nonlinear"
    return "ode"
end

"""
    declared_system_kind_mismatch(model::Model) -> Union{Tuple{String,String},Nothing}

`(declared, derived)` when the model declares a `system_kind` that contradicts
[`system_kind`](@ref), and `nothing` otherwise (including when the field is
absent). The `system_kind_mismatch` diagnostic reports the pair.
"""
function declared_system_kind_mismatch(model::Model)
    declared = model.system_kind
    declared === nothing && return nothing
    derived = system_kind(model)
    return declared == derived ? nothing : (declared, derived)
end

# ── Partition assertions (§6.3.1: "these sets PARTITION") ───────────────────

"""
    assert_classification_partitions(model::Model)

Check that the three unknown sets and the four parameter sets each partition
their side of the model's variables, throwing `ArgumentError` otherwise. The
spec states the partition property normatively, so it is asserted rather than
assumed; the conformance runner and `test/classification_test.jl` call it on
every fixture.
"""
function assert_classification_partitions(model::Model)
    us = ode_states(model)
    uo = observed_unknowns(model)
    ua = algebraic_unknowns(model)
    all_u = unknown_names(model)
    covered = vcat(us, uo, ua)
    length(covered) == length(unique(covered)) ||
        throw(ArgumentError("unknown sets overlap: $(sort!(covered))"))
    sort(covered) == all_u ||
        throw(ArgumentError("unknown sets do not cover the unknowns: " *
                            "$(sort!(covered)) vs $(all_u)"))

    pb = brownian_parameters(model)
    pd = discrete_parameters(model)
    ps = sampled_parameters(model)
    pc = constant_parameters(model)
    all_p = parameter_names(model)
    pcov = vcat(pb, pd, ps, pc)
    length(pcov) == length(unique(pcov)) ||
        throw(ArgumentError("parameter sets overlap: $(sort!(pcov))"))
    sort(pcov) == all_p ||
        throw(ArgumentError("parameter sets do not cover the parameters: " *
                            "$(sort!(pcov)) vs $(all_p)"))
    return nothing
end

File: pkg/earthsci-ast-go/pkg/esm/classification.go
package esm

import "sort"

// classification.go implements the esm-spec §6.3.1 classification API: the pure
// functions that recover, from the two declared variable types, the finer
// categories a solver needs.
//
// This file is the ONLY sanctioned way to ask these questions. Before 1.0.0 the
// format declared `state`, `observed`, `brownian` and `discrete`, and code
// branched on `variable.type == "state"`. Those types are gone, and reading a
// declared type to answer a derived question is exactly what 1.0.0 removes — so
// every such site calls in here instead.
//
// Two sets of PARTITIONS, pinned cross-language by
// tests/conformance/classification/:
//
//	unknowns   = ODEStates ⊎ ObservedUnknowns ⊎ AlgebraicUnknowns
//	parameters = BrownianParameters ⊎ DiscreteParameters ⊎
//	             SampledParameters ⊎ ConstantParameters
//
// Naming follows Go's idiom, as the spec's `naming` clause directs: the
// snake_case `ode_states` of Julia/Python/Rust is spelled `ODEStates` here.
//
// Every returned list is sorted lexicographically by byte order. That is what
// makes the cross-language goldens order-independent, and since
// CONFORMANCE_SPEC §7.1.0 it is also the order a list-valued diagnostic detail
// reports — Go decodes `variables` into a plain map and cannot produce
// declaration order at all.
//
// ── The independent variable ────────────────────────────────────────────────
//
// "Under D(·, t)" needs to know what `t` is, and the document — not the model —
// declares it (`domain.independent_variable`). The bare functions therefore
// assume the "t" default, and each carries an `…In(file, model)` companion that
// threads the document's actual choice. The four parameter partitions need no
// such companion: a parameter's cadence is decided entirely by its own
// `distribution` and `update`.

// The derived classes, as reported by FlattenedSystem.Variables and by any
// caller that wants to name a class rather than test membership.
const (
	ClassODEState          = "ode_state"
	ClassObserved          = "observed"
	ClassAlgebraic         = "algebraic"
	ClassBrownianParameter = "brownian_parameter"
	ClassDiscreteParameter = "discrete_parameter"
	ClassSampledParameter  = "sampled_parameter"
	ClassConstantParameter = "constant_parameter"
)

// ---------------------------------------------------------------------------
// Unknowns
// ---------------------------------------------------------------------------

// ODEStates returns the model's unknowns that appear under a time derivative
// `D(·, t)` on some equation LHS, sorted.
//
// The independent variable is assumed to be "t"; use ODEStatesIn for a document
// that declares another.
func ODEStates(model *Model) []string { return ODEStatesIn(nil, model) }

// ODEStatesIn is ODEStates against the document's declared independent variable.
func ODEStatesIn(file *ESMFile, model *Model) []string {
	return sortedSet(classifyUnknowns(file, model).ode)
}

// IsODEState reports whether name is one of the model's ODE states — the
// membership test §6.3.1 requires alongside the set.
//
// This is what a site that used to ask `variable.type == "state"` calls.
func IsODEState(model *Model, name string) bool { return IsODEStateIn(nil, model, name) }

// IsODEStateIn is IsODEState against the document's declared independent
// variable.
func IsODEStateIn(file *ESMFile, model *Model, name string) bool {
	return classifyUnknowns(file, model).ode[name]
}

// ObservedUnknowns returns the model's unknowns defined by a BARE-VARIABLE LHS
// (`y ~ f(…)`), sorted — the eliminable, materializable quantities.
//
// This is what a site that used to ask `variable.type == "observed"` calls, and
// where an observed's defining expression now lives: 1.0.0 removed the variable
// `expression` field, so the definition is the equation whose LHS is this name.
func ObservedUnknowns(model *Model) []string { return ObservedUnknownsIn(nil, model) }

// ObservedUnknownsIn is ObservedUnknowns against the document's declared
// independent variable.
func ObservedUnknownsIn(file *ESMFile, model *Model) []string {
	return sortedSet(classifyUnknowns(file, model).observed)
}

// AlgebraicUnknowns returns the model's unknowns constrained only IMPLICITLY —
// by an equation whose LHS is an arbitrary expression (`H*H*SO4 ~ Ksp`) rather
// than an assignment target — sorted.
//
// It is the residue of the partition: an unknown that is neither an ODE state
// nor observed. That includes an unknown NO equation mentions, which is an
// unbalanced system and is separately reported as equation_count_mismatch
// (esm-spec §4.9.4); classification stays total rather than silently dropping it
// and breaking the partition.
func AlgebraicUnknowns(model *Model) []string { return AlgebraicUnknownsIn(nil, model) }

// AlgebraicUnknownsIn is AlgebraicUnknowns against the document's declared
// independent variable.
func AlgebraicUnknownsIn(file *ESMFile, model *Model) []string {
	return sortedSet(classifyUnknowns(file, model).algebraic)
}

// Unknowns returns every unknown the model declares, sorted — the union the
// three partitions cover.
func Unknowns(model *Model) []string {
	out := map[string]bool{}
	if model != nil {
		for name, v := range model.Variables {
			if v.Type == VarTypeUnknown {
				out[name] = true
			}
		}
	}
	return sortedSet(out)
}

// ObservedDefinition returns the RHS of the equation defining an observed
// unknown, and whether one was found.
//
// This is the 1.0.0 relocation made available to callers: what used to be
// `variables[v].expression` is now the RHS of the bare-variable-LHS equation
// whose target is v. Any consumer that used to read the removed field asks here
// instead — including a cadence pass seeding an observed leaf from its defining
// equation (CONFORMANCE_SPEC §5.7.2), should this binding ever grow one.
func ObservedDefinition(model *Model, name string) (Expression, bool) {
	return ObservedDefinitionIn(nil, model, name)
}

// ObservedDefinitionIn is ObservedDefinition against the document's declared
// independent variable.
func ObservedDefinitionIn(file *ESMFile, model *Model, name string) (Expression, bool) {
	if model == nil {
		return nil, false
	}
	indep := indepVarOf(file)
	for _, eq := range model.Equations {
		if len(countDerivatives(eq.LHS, indep)) > 0 {
			continue
		}
		if assignmentTargetOf(eq.LHS) == name {
			return eq.RHS, true
		}
	}
	return nil, false
}

// unknownPartition is the computed three-way split of a model's unknowns.
type unknownPartition struct {
	ode       map[string]bool
	observed  map[string]bool
	algebraic map[string]bool
}

// classifyUnknowns performs the §6.3.1 unknown split in one pass.
//
// Precedence is ODE state ▸ observed ▸ algebraic, and it has to be a
// precedence rather than three independent tests, because the three sets must
// PARTITION: a variable both differentiated in one equation and assigned in
// another would otherwise land in two buckets. Differentiation wins — the
// solver integrates it either way, and the assignment is then a constraint on
// an integrated state rather than a definition of a derived one.
func classifyUnknowns(file *ESMFile, model *Model) unknownPartition {
	p := unknownPartition{
		ode:       map[string]bool{},
		observed:  map[string]bool{},
		algebraic: map[string]bool{},
	}
	if model == nil {
		return p
	}
	indep := indepVarOf(file)

	unknowns := map[string]bool{}
	for name, v := range model.Variables {
		if v.Type == VarTypeUnknown {
			unknowns[name] = true
		}
	}

	assigned := map[string]bool{}
	for _, eq := range model.Equations {
		// A derivative may be WRAPPED: `D(u)`, `D(u[i])` (i.e. `D(index(u,i))`),
		// and an `aggregate` whose `expr` is a `D(...)` all credit the base
		// variable. countDerivatives walks every expression-bearing field, so the
		// array-form equation that hides its D inside an aggregate's contracted
		// body is found too — looking only at a top-level `D` LHS is what used to
		// make Go report "0 ODE equations" for two dozen good fixtures.
		if derivs := countDerivatives(eq.LHS, indep); len(derivs) > 0 {
			for name := range derivs {
				if unknowns[name] {
					p.ode[name] = true
				}
			}
			continue
		}
		// Not a derivative: an assignment target makes the unknown OBSERVED, an
		// arbitrary expression LHS leaves it (and anything it mentions) algebraic.
		if target := assignmentTargetOf(eq.LHS); target != "" && unknowns[target] {
			assigned[target] = true
		}
	}

	for name := range unknowns {
		switch {
		case p.ode[name]:
			// already placed
		case assigned[name]:
			p.observed[name] = true
		default:
			p.algebraic[name] = true
		}
	}
	return p
}

// assignmentTargetOf returns the single variable an equation LHS assigns to — a
// bare name, an `index(v, …)` element, or the target inside an `aggregate`'s
// contracted body — and "" for an arbitrary expression LHS such as
// `H*H*SO4`, which assigns to nothing and therefore constrains implicitly.
//
// An `ic` LHS prescribes an initial value rather than the dynamics and is
// deliberately excluded: extractVariableFromLHS does not treat `ic` as an
// assignment target, so an `ic`-only unknown stays algebraic here and shows up
// in the balance check, which is where a missing defining equation belongs.
func assignmentTargetOf(lhs Expression) string { return extractVariableFromLHS(lhs) }

// ---------------------------------------------------------------------------
// Parameters
// ---------------------------------------------------------------------------

// BrownianParameters returns the model's parameters whose update kind is
// `wiener` — the SDE noise sources — sorted.
//
// This is what a site that used to ask `variable.type == "brownian"` calls. The
// 0.x `noise_kind` / `correlation_group` sidecars are gone: `wiener` is the only
// process, and correlated noise is one vector-valued parameter whose
// distribution carries a `cov`.
func BrownianParameters(model *Model) []string {
	return sortedSet(classifyParameters(model).brownian)
}

// DiscreteParameters returns the model's parameters carrying any OTHER update —
// piecewise-constant between refreshes — sorted.
//
// This is what a site that used to ask `variable.type == "discrete"` calls, and
// it is the sole seed of the DISCRETE cadence class (CONFORMANCE_SPEC §5.7.2).
func DiscreteParameters(model *Model) []string {
	return sortedSet(classifyParameters(model).discrete)
}

// SampledParameters returns the model's parameters with a `distribution` and NO
// `update` — drawn once at setup, the uncertainty-quantification / ensemble
// case — sorted.
func SampledParameters(model *Model) []string {
	return sortedSet(classifyParameters(model).sampled)
}

// ConstantParameters returns the model's parameters with neither a
// `distribution` nor an `update` — plain constants — sorted.
func ConstantParameters(model *Model) []string {
	return sortedSet(classifyParameters(model).constant)
}

// parameterPartition is the computed four-way split of a model's parameters.
type parameterPartition struct {
	brownian map[string]bool
	discrete map[string]bool
	sampled  map[string]bool
	constant map[string]bool
}

// classifyParameters performs the §6.3.1 parameter split in one pass.
//
// The two easy mistakes the conformance fixture `parameter_cadences` exists to
// catch are both avoided by testing UPDATE FIRST and DISTRIBUTION SECOND: a
// parameter with a distribution is not thereby Brownian (only a `wiener` update
// makes it so), and a parameter with an update is not thereby Brownian either
// (any non-wiener update makes it discrete, whether or not it also carries a
// distribution).
func classifyParameters(model *Model) parameterPartition {
	p := parameterPartition{
		brownian: map[string]bool{},
		discrete: map[string]bool{},
		sampled:  map[string]bool{},
		constant: map[string]bool{},
	}
	if model == nil {
		return p
	}
	for name, v := range model.Variables {
		if v.Type != VarTypeParameter {
			continue
		}
		switch {
		case v.Update.IsWiener():
			p.brownian[name] = true
		case v.Update != nil && len(v.Update.Rules) > 0:
			p.discrete[name] = true
		case v.Distribution != nil:
			p.sampled[name] = true
		default:
			p.constant[name] = true
		}
	}
	return p
}

// ClassOf returns the derived class of a declared variable — one of the seven
// Class* constants — and whether the model declares it at all. It is the
// single-name form of the two partitions, for a caller that wants to label a
// variable rather than enumerate a category.
func ClassOf(file *ESMFile, model *Model, name string) (string, bool) {
	if model == nil {
		return "", false
	}
	v, ok := model.Variables[name]
	if !ok {
		return "", false
	}
	if v.Type == VarTypeParameter {
		p := classifyParameters(model)
		switch {
		case p.brownian[name]:
			return ClassBrownianParameter, true
		case p.discrete[name]:
			return ClassDiscreteParameter, true
		case p.sampled[name]:
			return ClassSampledParameter, true
		default:
			return ClassConstantParameter, true
		}
	}
	u := classifyUnknowns(file, model)
	switch {
	case u.ode[name]:
		return ClassODEState, true
	case u.observed[name]:
		return ClassObserved, true
	default:
		return ClassAlgebraic, true
	}
}

// ---------------------------------------------------------------------------
// System kind
// ---------------------------------------------------------------------------

// SystemKind derives what the `system_kind` field declares (esm-spec §6.3.1),
// testing four conditions IN THIS ORDER and taking the first that holds:
//
//  1. any parameter in BrownianParameters       ⇒ "sde"
//  2. any equation contains a spatial derivative ⇒ "pde"
//  3. no time-derivative equation at all         ⇒ "nonlinear"
//  4. otherwise                                  ⇒ "ode"
//
// A binding uses the derivation when the field is ABSENT, and a present field
// that contradicts it is `system_kind_mismatch`.
//
// Both orderings are load-bearing and neither is interchangeable:
//
//   - "pde" BEFORE "nonlinear", because a steady-state PDE (`laplacian(u) ~ f`)
//     has no time-derivative equation and would otherwise fall through to
//     "nonlinear". It is a PDE, and PDESystem is what it maps to.
//   - "sde" BEFORE "pde", because a model carrying both a wiener parameter and a
//     spatial derivative has no SPDESystem constructor to select; it is
//     assembled as an SDE.
//
// The rule is stated over the EQUATIONS, not over a domain block: v0.8.0 removed
// `Domain.spatial`, so there is no spatial domain left to test and the
// differential operators are the whole signal.
func SystemKind(model *Model) string { return SystemKindIn(nil, model) }

// SystemKindIn is SystemKind against the document's declared independent
// variable.
func SystemKindIn(file *ESMFile, model *Model) string {
	if model == nil {
		return SystemKindODE
	}
	if len(BrownianParameters(model)) > 0 {
		return SystemKindSDE
	}
	indep := indepVarOf(file)
	if hasSpatialDerivative(model, indep) {
		return SystemKindPDE
	}
	for _, eq := range model.Equations {
		if len(countDerivatives(eq.LHS, indep)) > 0 {
			return SystemKindODE
		}
	}
	return SystemKindNonlinear
}

// DeclaredSystemKind returns the model's explicit `system_kind` field and
// whether it was present — the `declared_system_kind` of the classification
// goldens, and the left-hand side of a system_kind_mismatch comparison.
func DeclaredSystemKind(model *Model) (string, bool) {
	if model == nil || model.SystemKind == nil {
		return "", false
	}
	return *model.SystemKind, true
}

// spatialSugarOps are the sugar spellings of a spatial derivative (esm-spec
// §4.2). They are OPEN-tier rewrite targets, so neither they nor the explicit
// `D`-with-a-spatial-`wrt` form is canonical and both must be recognized.
var spatialSugarOps = map[string]bool{
	"grad": true, "div": true, "laplacian": true,
}

// hasSpatialDerivative reports whether any equation of the model contains a
// spatial derivative: a `D` node whose `wrt` is PRESENT and is not the
// independent variable, or one of the grad / div / laplacian sugar ops.
//
// It is detected anywhere in an equation's LHS or RHS, and the walk descends
// every Expression child rather than `args` alone (§4.9.5) — a `laplacian`
// buried in an aggregate's `expr` or an integral bound is still a spatial
// derivative.
//
// A `D` with NO `wrt` is a time derivative by the package-wide convention
// (isDifferentialEquation, countDerivatives), so it is deliberately not spatial
// here.
func hasSpatialDerivative(model *Model, indep string) bool {
	found := false
	var walk func(Expression)
	walk = func(e Expression) {
		if found {
			return
		}
		node, ok := asExprNode(e)
		if !ok {
			return
		}
		if spatialSugarOps[node.Op] {
			found = true
			return
		}
		if node.Op == OpDerivative && node.Wrt != nil && *node.Wrt != indep {
			found = true
			return
		}
		_, _ = mapExprChildren(node, func(child Expression) (Expression, error) {
			walk(child)
			return child, nil
		})
	}
	for _, eq := range model.Equations {
		walk(eq.LHS)
		walk(eq.RHS)
		if found {
			return true
		}
	}
	return false
}

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

// indepVarOf returns the document's independent variable, defaulting to "t"
// when there is no document or it declares none. It is fileIndepVar widened to
// accept a nil file, so the bare (model-only) classification functions and their
// `…In` companions share one code path.
func indepVarOf(file *ESMFile) string {
	if file == nil {
		return DefaultIndepVar
	}
	return fileIndepVar(file)
}

// sortedSet renders a membership set as a lexicographically sorted slice. A nil
// or empty set renders as an EMPTY slice rather than nil, so it marshals to `[]`
// and not `null` — the classification goldens spell an empty category `[]`.
func sortedSet(set map[string]bool) []string {
	out := make([]string, 0, len(set))
	for name := range set {
		out = append(out, name)
	}
	sort.Strings(out)
	return out
}
